lib/SILOptimizer/Transforms/ARCCodeMotion.cpp

```diff
@@ -375,16 +377,50 @@ bool RetainCodeMotionContext::requireIteration() {
 
 void RetainCodeMotionContext::initializeCodeMotionDataFlow() {
   // Find all the RC roots in the function.
+  llvm::DenseMap<SILValue, Atomicity> AtomicityOfRoot;
+
   for (auto &BB : *F) {
     for (auto &II : BB) {
+      auto RCInst = dyn_cast<RefCountingInst>(&II);
+      if (!RCInst)
+        continue;
+      // Remember the atomicity of the root. If any
+      // of the RC operations on a the root are atomic,
+      // then the root is considered to be atomic.
+      Atomicity InstAtomicity = (RCInst->getAtomicity() == Atomicity::Atomic)
+                                    ? Atomicity::Atomic
+                                    : Atomicity::NonAtomic;
+      //auto Root = getRCRoot(&II);
+      auto Root = getRCRoot(RCInst->getOperand(0));
+
+      auto newRoot = AtomicityOfRoot.count(Root);
+      auto &RootAtomicity = AtomicityOfRoot[Root];
+
+      if (!newRoot)
```

Comment:
@swiftix I think one way to do this is to force an ordering on the 2 passes. i.e. we can not run ARCCodeMotion (ACM) after the ARC atomic pass. This is not perfect, but some of our passes do have ordering. My feeling is to have ARC atomic pass should run after all the ARC passes, ASO, ACM are done running.